### ISSUE TITLE
Is updating to uBlock Origin 1.17.x possible?

### ISSUE THREAD

**Issue by palopezv** [nikrolls/uBlock-Edge#122]
As the title says.

**Comment by Weasy666** [comment]
I'd also like to see an update. There is this annoying problem with the entire browser getting minimized when you deactivate ublock on a page and reload it from the ublock menu.

**Comment by jspenguin2017** [comment]
@gorhill Edge port is stuck (again), this time on 1.15.24. 

I think the problem is that nikrolls rewrote `vAPI` scripts, which means every time upstream changes it's a huge pain to pull the changes. 
I instead used `-ms-preload` to insert a shim, and the unshimmable changes is only a few lines: https://github.com/NanoAdblocker/NanoCore2/blob/master/patches/integration/90_0_edgyfy.patch

Except the `tabs.reload` function, all works fine on `1.16.21rc0`.

**Comment by nikrolls** [comment]
There has been very little rewritten. Only small modifications have been made where there are incompatibilites with Edge.

The issue at the moment is that recent releases of uBlock are broken in Edge with regards to hiding elements. This is to do with the CSS content rewrites in the core. I'm monitoring it and trying new releases as they come out, as well as attempting to find a fix.

I welcome any and all PRs. This is a mostly solo effort so far so I'd prefer support over complaints.

**Comment by jspenguin2017** [comment]
@nikrolls Try this patch? https://github.com/NanoAdblocker/NanoCore2/blob/master/patches/10_0_user_css.patch

**Comment by gorhill** [comment]
Whatever can go to upstream is welcome, I already have code paths specific to Firefox or Chromium, I don't mind also doing it for Edge -- so as to minimize as much as possible Edge version.

**Comment by jspenguin2017** [comment]
@gorhill I'm using the shim here: https://github.com/jspenguin2017/Edgyfy/blob/master/edgyfy.js
No other changes should be needed beside the two patches mentioned above. 

The maintenance work is basically 0. It usually takes 2~3 minutes per release. I would say just put Edge stuff in Chromium platform scripts, 1 hash table lookup really does not slow things down. 

There's also a browser bug (`tabs.reload`, and `tags.select`) that I'm planning to report. It's not show stopping but a bit annoying.

**Comment by jspenguin2017** [comment]
@gorhill Also keep in mind that some filters needs to be different in Edge, and neither uAssets nor the Subreddit are interested in handling them.

**Comment by gorhill** [comment]
I can't test nothing for the Edge version -- I don't have Windows, so best is that a PR is made with everything sorted out already, the only thing I will be able to look at is that both Chromium/Firefox are still working as expected.

**Comment by jspenguin2017** [comment]
@gorhill You can run a VM, it's free, Windows 10 doesn't force you to buy a license.

**Comment by gorhill** [comment]
@nikrolls I have created a new branch for Edge, <https://github.com/gorhill/uBlock/tree/edge>, which is essentially the chromium version + all changes I found was needed to remove errors and glitches that I could spot when loading uBO in Edge (from within a VM). Surely there is more work to do but I think it would be best now to bring the Edge version in the main dev tree since there are really not many changes required anymore. I've added you as collaborator so you can change whatever is needed without having to go through pull requests.

**Comment by jspenguin2017** [comment]
@gorhill Do you have access to the store listing? Also make sure you have Windows SDK installed, otherwise ManifoldJS will send your code to a build server over HTTP. 

I should be able to help with testing by the end of the month.

**Comment by gorhill** [comment]
No, I don't have access to the listing. At this point I am waiting an answer from @nikrolls in order to proceed with the above.

**Comment by jspenguin2017** [comment]
@gorhill Also `ms-browser-extension-scheme` should be in default whitelist.

**Comment by jspenguin2017** [comment]
And you should sanitize `Date.prototype.toLocaleString` if you haven't: 
![image](https://user-images.githubusercontent.com/7283682/45921151-67fac300-be6d-11e8-9049-ed7bc0209ca1.png)

**Comment by James1o1o** [comment]
Pinging @nikrolls

**Comment by jspenguin2017** [comment]
Maybe we should just start fresh, clearly this isn't going anywhere. 

I think my current patch driven approach is working way better than a normal fork. I used to use a normal fork, and 4 months in it's basically impossible to pull upstream changes. 

Running 1.17.0 on 1803 without any showstopping issues: 

<details>

![image](https://user-images.githubusercontent.com/7283682/48090512-61799e00-e1c4-11e8-97de-57d8201f9306.png)
![image](https://user-images.githubusercontent.com/7283682/48090519-68a0ac00-e1c4-11e8-8f68-e29d78160c6d.png)

</details>

**Comment by Weasy666** [comment]
Has anyone created a new repo? Or is it possible to integrate it somehow with what @gorhill proposed?

**Comment by jspenguin2017** [comment]
I guess it doesn't matter anymore, Edge will soon be based on Chromium.

**Comment by palopezv** [comment]
If you consider October 2019, if we are lucky, close enough. Judging by Microsoft's development track record I would guess October 2020 is a better possibility.

**Comment by jspenguin2017** [comment]
Chromium is a complete and functional browser, all Microsoft has to do is to make it look more like the current Edge, I'd say it'll be ready next April.

**Comment by nikrolls** [comment]
Apologies again for my absense, domestic issues are still taking much more of my time than I expected.

[Microsoft has announced that they intend Edge on Chromium to "support existing Chrome extensions" (Engadget)](https://www.engadget.com/2018/12/10/microsoft-edge-chrome-extensions/). Will that mean it can literally access the Chrome Web Store? I think unlikely. Will it mean that you can submit the exact same package to the Chrome Web Store and the Microsoft Store? At the very least. 

As I've stated earlier, I will be fully cooperative with regards to transferring the Microsoft Store application or name (whatever is required) to @gorhill or someone he nominates if desired. I am passionate about uBlock on Edge and always intended to continue to maintain it, but I've learned that when bad thing happen in life it can be very hard to anticipate how much of a time suck they can be or when you will fully recover from them.

**Comment by jspenguin2017** [comment]
> I think unlikely. 

I think very likely. I would guess 99% of developers would not be interested in spending $20 just to copy the exact same package from WebStore to Microsoft Store, so it does not make sense for Microsoft to restrict extension pool of Edge to just Microsoft Store. Considering other Chromium forks are able to install extensions from WebStore, I don't see why Microsoft would not do the same. 

> transferring the Microsoft Store application or name

I'd say it's better if @gorhill can hold the Microsoft Store listing, since Edge is going to be a fork of Chromium, there is really no need to nominate someone just to publish the package. As for the gap between now and the new Edge release, I don't think anyone would want to spend time debugging it just to dump everything a few months later.

**Comment by gorhill** [comment]
> I will be fully cooperative with regards to transferring the Microsoft Store application or name

Well given that Edge will be based on Chromium in the medium term, there is no point to keep working on the current Edge version. However, once Edge uses Chromium engine, I expect that uBO will work with minimal changes (maybe details in the manifest), so that should lower quite a lot maintenance burden. I guess it's a few months in the future now, but once Edge converts to Chromium engine, is there a possibility that you will be able again to take care of uBO/Edge version?

**Comment by jspenguin2017** [comment]
I just got an email from Microsoft about the new Edge browser. They claim that Chromium extensions should work without modification. However, it sounds like the new Edge browser won't be able to use the WebStore, but that's not explicitly mentioned in the email, so I could be wrong. Microsoft will test and publish existing extensions if listing owners provided consent. 

Looks like the new Edge browser is just around the corner.

**Comment by nikrolls** [comment]
Yep, looks like Microsoft will maintain its own store (which while less than ideal for users, makes sense for a lot of reasons). I can't tell if they will automatically publish extensions by scraping updates from the Chrome store, but they seem to want to do that with some extensions during testing.

@gorhill I will contact you privately about getting uBlock into this testing programme, and to your other question yes I would be happy to maintain this process in the future if Microsoft don't automate it.

**Comment by superusercode** [comment]
https://blogs.windows.com/msedgedev/2019/12/16/get-started-extensions-addons-microsoft-edge-chromium/

## Transitioning your existing extensions to Chromium
As we move towards the general availability of the new Microsoft Edge on January 15th, 2020, we will no longer accept new submissions for Microsoft Edge Legacy (EdgeHTML-based) extensions after December 17th, 2019. We will continue to accept updates for your existing extensions.

We recommend you prioritize building new extensions for the new Chromium-based Microsoft Edge, and continue to support your existing EdgeHTML-based extensions to ensure a quality experience for active users.

Developers who have given consent for Microsoft to migrate their EdgeHTML extension listings to the new Microsoft Edge should begin to see their extensions available in the new Addons store experience in Microsoft Edge. If you publish an EdgeHTML extension and have not received any communication regarding its migration or are unsure of its status, please contact us at ExtensionPartnerOps@microsoft.com.

If you have already received a confirmation from us regarding migration, we encourage you to log on to the Partner Center Developer Dashboard to validate your access to the extension, and verify whether you can update it. Once the migration is complete, ownership and management will be completely transferred to you, and **Microsoft will not be responsible for updating or maintaining your extension.**
